### PR TITLE
Testing WebDriverIO

### PR DESCRIPTION
**Pull request by ordavidil**
#29

### CODE REVIEW

File: ci-scripts/docker_files/preparing_client.sh [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+cd /opt/
+wget --no-cookies --no-check-certificate --header "Cookie: gpw_e24=http%3A%2F%2Fwww.oracle.com%2F; oraclelicense=accept-securebackup-cookie" "http://download.oracle.com/otn-pub/java/jdk/7u79-b15/jdk-7u79-linux-x64.tar.gz"
+tar xzf jdk-7u79-linux-x64.tar.gz
+
+# Install Chrome driver.
+wget https://chromedriver.storage.googleapis.com/2.27/chromedriver_linux64.zip | tar xzf
+
+apt-get install openjdk-7-jre-headless -y
+
+# Install Selenium.
```
Comment (amitaibu) [inline ci-scripts/docker_files/preparing_client.sh]: selenium isn't needed for the tests, as they run via Chrome driver directly.

**Comment by amitaibu** [comment]
@ordavidil maybe instead of trying to install Chrome, you can use another docker container for this. [See](http://www.gizra.com/content/phantomjs-chrome-docker-selenium-standalone/), which is using https://github.com/elgalu/docker-selenium. Otherwise, maybe that repo can at least act as a source to see how do it.

**Comment by ordavidil** [comment]
@ItamarGronich 
Check this PR ^^

**Comment by AronNovak** [comment]
@ordavidil It seems i can start it now, not sure if i will have a clean timeslot now, but just took a look.
So
```
 > Test WebDriverIO. 
[07:31:28] Using gulpfile /var/www/html/Server/client/gulpfile.js

ERROR: Couldn't initialise service "phantomjs".
Error: Cannot find module 'wdio-phantomjs-service'

...
   at process.<anonymous> (/var/www/html/Server/client/node_modules/webdriverio/build/lib/runner.js:749:22)
...
The command "$TRAVIS_BUILD_DIR/ci-scripts/test_server.sh" exited with 1.
```

What's on those lines from the build output, it sums up the issue nicely, right?

**Comment by amitaibu** [comment]
when chrome driver isn't running, it's falling back to phantomjs

**Comment by AronNovak** [comment]
Ok, thanks. Right now I try to see the very same fail at localhost:

```
 aaron   29-webdriverio-travis  ⋯  drupal-elm-starter  ci-scripts  docker_files  docker build -t "gizra:drupal-elm-starter" .
Sending build context to Docker daemon 14.85 kB
Step 0 : FROM gizra/drupal-lamp
 ---> 0a63954adc6f
Step 1 : ADD $TRAVIS_BUILD_DIR /var/www/html/Server
 ---> Using cache
 ---> 5b663a5c8236
Step 2 : WORKDIR /var/www/html/Server
 ---> Using cache
 ---> 812c0137bbf2
Step 3 : RUN chmod +x /var/www/html/Server/ci-scripts/docker_files/run.sh
 ---> Running in 544d5945b694
chmod: cannot access '/var/www/html/Server/ci-scripts/docker_files/run.sh': No such file or directory
The command '/bin/sh -c chmod +x /var/www/html/Server/ci-scripts/docker_files/run.sh' returned a non-zero code: 1
```

But it fails to build an image from the Dockerfile. Any hints? Thanks!

**Comment by ordavidil** [comment]
@AronNovak 
what's the command you invoke in order to build the image?
and what's the command to run the container ?

**Comment by AronNovak** [comment]
I am at the point just to build the image, the command is: `docker build -t "gizra:drupal-elm-starter" .`
And i am in `/home/aaron/gizra/drupal-elm-starter/ci-scripts/docker_files`

**Comment by ordavidil** [comment]
Build the image:
```sh
docker build -t server -f ci-scripts/docker_files/Dockerfile .
```

Run the container:
```sh
docker run -it -p 8080:80 -e BUILD_WEBDRIVERIO=1 server
```

**Comment by amitaibu** [comment]
@AronNovak ^^ I think worth adding this info to the README under `/ci-scripts`

**Comment by AronNovak** [comment]
The commit to add the README.md is ready, but:
```ERROR: Permission to Gizra/drupal-elm-starter.git denied to AronNovak.
fatal: Could not read from remote repository.
```

**Comment by amitaibu** [comment]
I've granted you write access

**Comment by AronNovak** [comment]
Thanks! Push and local test execution works like a charm.
Now I study https://docs.docker.com/compose/overview/ and https://hub.docker.com/r/huli/webdriverio/

**Comment by AronNovak** [comment]
```
docker ps
CONTAINER ID        IMAGE                  COMMAND                  CREATED             STATUS              PORTS                                    NAMES    
373c6955b1f4        server                 "/var/www/html/Ser..."   2 minutes ago       Up 2 minutes        22/tcp, 3306/tcp, 0.0.0.0:8080->80/tcp   fervent_euler                                                                                                                                                          
de6ad6070549        selenium/node-chrome   "/opt/bin/entry_po..."   7 minutes ago       Up 7 minutes        0.0.0.0:32768->5900/tcp                  dockerfiles_chrome_1                                                                                                                                                   
2fef0b3b980b        selenium/hub           "/opt/bin/entry_po..."   7 minutes ago       Up 7 minutes        0.0.0.0:4444->4444/tcp                   dockerfiles_hub_1
```

`huli/webdriverio` is up and running, checking how to connect it to our original container

**Comment by AronNovak** [comment]
Hm, after I let `huli/webdriverio`, the original docker container loses the network connection, error message appears like as:
https://github.com/docker/docker/issues/5618
https://github.com/coreos/bugs/issues/254

I'll check if `docker-compose` is supposed to launch the server image too or not, maybe that changes the things.

**Comment by AronNovak** [comment]
I have a generic Docker issue almost, it seems. Docker-compose required a newer version of Docker (the one used was Fedora 24 bulilt-in). After upgrading, from the `server`, networking stopped.
However:
```
docker run --rm busybox traceroute gizra.com
```

This one does not fail. So it's not that networking is fully broken under Docker after the upgrade, but the server image cannot fetch anything external (rebuilt the image after the upgrade)

**Comment by AronNovak** [comment]
I could solve all the network issues by this config:
```
~  gizra  drupal-elm-starter  cat /etc/systemd/system/docker.service | grep mtu
ExecStart=/usr/bin/dockerd --mtu=1450
```
Now I can proceed with the actual issue...

**Comment by AronNovak** [comment]
In theory, I have everything to make WDIO.js on server to  recognize the other container, all I need to do is to execute this:
```
docker-compose up
```
But it does not get to the point where the execution would start, because:
```
server_1       | [10:54:37] Starting 'copy:dev'...
server_1       | [10:54:37] index.html & CNAME all files 1.91 kB
server_1       | [10:54:37] Finished 'copy:dev' after 15 ms
server_1       | ## There is an issue with `node-fibers` ##
server_1       | `/var/www/html/Server/client/node_modules/wdio-mocha-framework/node_modules/wdio-sync/node_modules/fibers/bin/linux-x64-47/fibers.node` is missing.
server_1       | 
server_1       | Try running this to fix the issue: /usr/bin/nodejs /var/www/html/Server/client/node_modules/wdio-mocha-framework/node_modules/wdio-sync/node_modules/fibers/build
server_1       | ERROR: Couldn't initialise framework "wdio-mocha-framework".
```

It seems fully unrelated to our current PR, as it fails at Gulp. Any hints how I can solve this quickly?

**Comment by amitaibu** [comment]
maybe find which module in `package.json` causes it, and see if upgrading the version could help

**Comment by amitaibu** [comment]
Or trying to follow the error message: `/usr/bin/nodejs /var/www/html/Server/client/node_modules/wdio-mocha-framework/node_modules/wdio-sync/node_modules/fibers/build`?

**Comment by AronNovak** [comment]
`/usr/bin/nodejs /var/www/html/Server/client/node_modules/wdio-mocha-framework/node_modules/wdio-sync/node_modules/fibers/build` this alone did not help, i tried to proceed https://github.com/laverdet/node-fibers/issues/331, now i don't have a fibers issue, but there's an issue with phantomjs. I nearly exhausted the timebox, for now the issue is this processing of `package.json` that's not completed successfully:
```
npm ERR! Linux 4.9.10-100.fc24.x86_64
npm ERR! argv "/usr/bin/nodejs" "/usr/bin/npm" "install"
npm ERR! node v5.12.0
npm ERR! npm  v3.8.6
npm ERR! code ELIFECYCLE

npm ERR! phantomjs@1.9.20 install: `node install.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the phantomjs@1.9.20 install script 'node install.js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the phantomjs package,
npm ERR! not with npm itself.
```

Despite the fact that I removed temporarily the phantomjs-packages from `client/package.json` (and rebuilt the image).
May I proceed further with debugging this?

**Comment by AronNovak** [comment]
Now I understand, by comparing to `master` `ci-scripts`; that previously we never built the whole client, all we settled was Elm, so this issue also includes a full-blown client installation, not just the integration of WDIO itself.

**Comment by ordavidil** [comment]
Sorry if I wasn't clear when we talked.
Yes,
the main issues are:
1. build the client
2. run the gulp so the client will be served
3. run the WDIO tests.

**Comment by AronNovak** [comment]
Thanks for the feedback.
This morning's progress:
```
# Fibers Node 7.x issue: https://github.com/laverdet/node-fibers/issues/331
apt-get -y install g++-4.8
export CXX=g++-4.8

apt-get -y install bzip2
```

With these, the client dependencies with `npm install` could be completed successfully.

**Comment by AronNovak** [comment]
test-execution

Now we arrived to the point that:
* client is built successfully
* the test execution is started, the two containers can interact nicely

Next steps:
* check if the test failure is due to the internal Docker environment or not, compare it to an outside test execution
* possibly reduce the diff compared to `master` branch, remove experimental changes and see if things can work without those (namely `node.js` downgrade, `package.json` manipulation related to `PhantomJS`)

**Comment by amitaibu** [comment]
The failing tests might be related to the test itself - is it passing locally for you?

**Comment by ordavidil** [comment]
@AronNovak 
pay attention that the _Client_ build is now failed ...

**Comment by AronNovak** [comment]
@ordavidil  thanks for the heads-up, i tested the node.js version downgrade locally, as it did not fail at local Docker execution, I pushed, now client build is back.
I started to check if the tests pass locally or not, in my native environment.

**Comment by AronNovak** [comment]
@amitaibu The tests do pass locally:
passing
(after re-configuring of course a few trivial things in wdio.conf.js, like Selenium's hostname and local hostname, as i use my .local domains everywhere)

So it's still something with the environment, i will keep checking what. My plan is now to somehow prevent the Docker container to stop after the test execution, so i'd be able to do some experiment with it after a failed test. That's my biggest annoyance right now with the debugging, ideally I imagine, in debug mode, I could get a shell after the test execution, to be able to inspect.

**Comment by ordavidil** [comment]
@AronNovak 
we probably want to use `wdio.conf.js.travis` as usual

@amitaibu 
please confirm

**Comment by AronNovak** [comment]
Added the ability to be able to interactively debug the Docker container and documented it.
container-debug

I hope it will help to tackle the rest of this task faster.

**Comment by AronNovak** [comment]
The summary of current debubbing via Docker interactive Bash:
My assumption is that the test fails because the backend is not accessible via the same hostname `localhost`:
```
root@1abcc576bc7e:/var/www/html/Server/client# curl --silent http://localhost/ | tail
<html><head>
<title>403 Forbidden</title>
</head><body>
<h1>Forbidden</h1>
<p>You don't have permission to access /
on this server.<br />
</p>
<hr>
<address>Apache/2.4.10 (Debian) Server at localhost Port 80</address>
</body></html>
root@1abcc576bc7e:/var/www/html/Server/client# curl --silent http://server.local/ | tail
  
  <div id="footer-wrapper"><div class="section">
```

So likely if i reconfigure Apache, i will be one step further.

**Comment by ordavidil** [comment]
@AronNovak 
There should be already a vhost to the server `http://server.local`

**Comment by amitaibu** [comment]
> There should be already a vhost to the server http://server.local

@AronNovak if the client side is trying to reach the wrong server, you can change it by having a [local config](https://github.com/Gizra/drupal-elm-starter/pull/46) with the correct Backend URL

**Comment by AronNovak** [comment]
A quick update, I spent a little time to proceed here:
```
 aaron   29-webdriverio-travis  ⋯  drupal-elm-starter  ci-scripts  docker_files  gd server.conf
diff --git a/ci-scripts/docker_files/server.conf b/ci-scripts/docker_files/server.conf
index cf4e1e8..7926b6e 100644
--- a/ci-scripts/docker_files/server.conf
+++ b/ci-scripts/docker_files/server.conf
@@ -1,4 +1,4 @@
 <VirtualHost *:80>
     DocumentRoot /var/www/html/Server/server/www
-    ServerName server.local
+    ServerName server.local localhost
 </VirtualHost>
```
Did not solve the issue, try to check if indeed dockerized client can talk to dockerized server.

**Comment by AronNovak** [comment]
https://github.com/Gizra/drupal-elm-starter/pull/43#issuecomment-283382383 is a syntax error in Apache config, of course it did not solve anything, just made it worse. Now I was able to reach the backend from any path, inside, outside, firing `docker-compose up`

File: ci-scripts/docker_files/server.conf [hunk 2]
```diff
@@ -1,4 +1,5 @@
 <VirtualHost *:80>
     DocumentRoot /var/www/html/Server/server/www
     ServerName server.local
+    ServerAlias localhost
```
Comment (amitaibu) [inline ci-scripts/docker_files/server.conf]: remember that you can have your own `LocalConfig.elm` - so it will contact any backend that you want (because  now the default is `http://localhost/drupal-elm-starter/www`

Comment (AronNovak) [inline ci-scripts/docker_files/server.conf]: Yep, thanks! But i am bound to https://en.wikipedia.org/wiki/Same-origin_policy , not? So if it's not on the same hostname, it won't work. I did not do lots of headless, but my assumption is the hostname should be common, there are workarounds for that, but that's the simplest.

Comment (amitaibu) [inline ci-scripts/docker_files/server.conf]: Not at all, we are using an access token that can be sent from anywhere

Comment (AronNovak) [inline ci-scripts/docker_files/server.conf]: That's good to know, i proceed this way then, will make the things easier!

**Comment by AronNovak** [comment]
Now the client + server is fully sane inside the containers, if i map the ports, i can use from my local browser.
docker-env
The relevant command to launch it was:
```
docker run -it -p 80:80 -p 3000:3000 -e BUILD_WEBDRIVERIO=1 -e DOCKER_DEBUG=1 server
```

The remaining steps are:
* check why the container with Chrome cannot connect to the client's web port / hostname
* optionally check if tests are okay
* integrate the `docker-compose up` part with Travis

**Comment by AronNovak** [comment]
One step further:
```
seluser@0fe1d6207094:/opt/selenium$ java -jar selenium-server-standalone.jar --version
Selenium server version: 3.1.0, revision: 86a5d70
```
As i know our tests should work with 2.x Selenium not 3.x, trying to check how to downgrade easily, possibly with another pre-build container, from the suggested ones above.

**Comment by amitaibu** [comment]
we don't need selenium server to run, we just need chrome and chromdriver

**Comment by AronNovak** [comment]
docker

Now we arrived to the point that the test passes within Docker @amitaibu thanks for the tremendous help.From the previous commit, i did not need to change anything, the failure was due to a local packet filtering firewall config.

**Comment by AronNovak** [comment]
The current status is that we need to make sure that:
 * the tests pass under Travis
 * we efficiently execute the things, so we don't spin up unneeded selenium container, when not needed, etc.

**Comment by AronNovak** [comment]
Right now the Webdriver.io test execution is started at Travis, but it fails, with the same error as it was on my local environment, there's an issue which is likely related, as also at my local, cross-container communication was blocked:
* https://github.com/travis-ci/travis-ci/issues/4842
If it's not possible to workaround easily with configuration options, a way could be to change port numbers, people reported to have success with that.

**Comment by AronNovak** [comment]
The current status is unfortunately the same after today's commit:
 * the WDIO tests pass locally via docker-compose
 * the WDIO tests fail at Travis CI, with the same symptom (10000 ms timeout) as it failed at local with the firewall
 * some minor tweaks happened today related to Travis-based execution like exit status.

So still todo:
 * Make the tests pass on Travis (possibly related: https://github.com/travis-ci/travis-ci/issues/4842)
 * Make the test execution more efficient, do not execute unneeded steps to save time

**Comment by amitaibu** [comment]
@ordavidil please guide @AronNovak on how he can debug a Travis instance.

**Comment by amitaibu** [comment]
btw @AronNovak , note that I see other envs failing as-well, so worth fixing them

**Comment by amitaibu** [comment]
build__170_-_gizra_drupal-elm-starter_-_travis_ci

**Comment by ordavidil** [comment]
@amitaibu 
Since the button i smissing I've been trying to look for it but couldn't find anything related to it on the docs / github.

My assumption is that because the repo is public (can not think about any other reason)

**Comment by amitaibu** [comment]
Worth contacting them to ask

**Comment by AronNovak** [comment]
@amitaibu @ordavidil Thanks for the feedback, I go ahead and I contact Travis about this issue and going to make the other two builds pass in the meantime of course..

**Comment by AronNovak** [comment]
Mail sent to Travis support, and other 2 tests were fixed:
fail
On hold until received answer about the cross-container communication issue.

**Comment by amitaibu** [comment]
> On hold until received answer about the cross-container communication issue.

They probably won't know how to answer. The support should be about how to allow us to debug a Travis instance (they have a way of letting you to ssh into the box)

**Comment by AronNovak** [comment]
Indeed, thanks for the notice, then I just did a follow-up on the thread mentioning that debugging (SSH, etc) would be sufficient help for us.

**Comment by AronNovak** [comment]
The support was swift, they enabled us to debug the build with SSH, so hopefully we will be able to soon solve this, I tested the SSH access, it works like a charm, the only limitation is that only one SSH session is enabled, but there are workarounds for that, for instance `tmux`, as i will need to issue `docker` commands while docker-compose up is running.

**Comment by AronNovak** [comment]
We're one step further. one of the tests passes at Travis server:
pass-fail

There was a timing issue, at localhost, everything completed faster, at `Travis`, as we send `gulp` to the background, we started to test even before the application was up and running via `gulp`.

**Comment by amitaibu** [comment]
nice!

Usually, on travis we gulp and then `curl` the url until we get a 200

**Comment by amitaibu** [comment]
https://github.com/Gizra/gizra.com/blob/e4a73f5f32f7cc10feec0252a86ff2b44c670be2/.travis.yml#L30

**Comment by AronNovak** [comment]
Tests are passing, I am doing some improvements like https://github.com/Gizra/drupal-elm-starter/pull/43#issuecomment-284650629 and so on!

**Comment by AronNovak** [comment]
I performed the tweaks I had in mind, any further feedback is highly appreciated!

**Comment by amitaibu** [comment]
🎉 It's green!

File: client/src/elm/Pages/Login/Update.elm [hunk 3]
```diff
@@ -1,6 +1,7 @@
 module Pages.Login.Update exposing (fetchUserFromBackend, update)
 
 import Config.Model exposing (BackendUrl)
+import Http exposing (Error(BadStatus))
```
Comment (amitaibu) [inline client/src/elm/Pages/Login/Update.elm]: is this change related, or just need a re-roll with master?

Comment (AronNovak) [inline client/src/elm/Pages/Login/Update.elm]: It's coming from https://github.com/Gizra/drupal-elm-starter/pull/43/commits/c244a190b098107ce0eac601ec312915c4752e26 , on my side, it's not related. It seems, at a certain point, the Elm app/test itself was modified for some reason, we might revert those commits, if not needed.

Comment (AronNovak) [inline client/src/elm/Pages/Login/Update.elm]: @rohanorton could you help us about the purpose of the Elm-related commits under this PR? Thank you!

Comment (rohanorton) [inline client/src/elm/Pages/Login/Update.elm]: They're from this PR: https://github.com/Gizra/drupal-elm-starter/pull/51. It was to prevent error messages being shown on invalid session tokens.

**Comment by amitaibu** [comment]
Sorry, what I meant was that you had to `git merge master` into the branch. I've done so, and waiting for green to merge.

Amazing work folks!

File: client/wdio.conf.js [hunk 4]
```diff
@@ -72,10 +72,10 @@ exports.config = {
     //
     // Set a base URL in order to shorten url command calls. If your url parameter starts
     // with "/", then the base url gets prepended.
-    baseUrl: 'http://localhost:3000',
+    baseUrl: 'http://server.local:3000',
```
Comment (amitaibu) [inline client/wdio.conf.js]: @AronNovak oh, this was changed on the general `client/wdio.conf.js`. Please create a `client/wdio.conf.travis.js` and execute it instead, and revert the changes introduced here
